File: week1/1.two-sum/1.two-sum.js
// O(n)
var twoSum = function (nums, target) {
    let map = new Map();
    for (let i = 0; i < nums.length; i++) {
        const v = nums[i];
        if (map.has(target - v)) {
            const index = map.get(target - v)
            return [i, index]
        }
        map.set(v, i)
    }
    return [null, null];
};

// O(n^2)
var twoSum = function (nums, target) {
    for (let i = 0; i < nums.length - 1; i++) {
        for (let j = i + 1; j < nums.length; j++) {
            if (nums[i] + nums[j] == target) {
                return [i, j];
            }
        }
    }
    return [null, null];
}

// O(nlogn) 
var twoSum = function (nums, target) {
    // sort
    // BS
}

File: week1/1.two-sum/1.two-sum.go
// O(n^2)
func twoSum(nums []int, target int) []int {
	var result = make([]int, 2)
	for i := 0; i < len(nums)-1; i++ {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				result[0], result[1] = i, j
			}
		}
	}
	return result
}

// O(n)
func twoSum(nums []int, target int) []int {
	hm := map[int]int{}
	for i := 0; i < len(nums); i++ {
		v := nums[i]
		if idx, ok := hm[target-v]; ok {
			return []int{idx, i}
		}
		hm[v] = i
	}
	return nil
}